compiler/rustc_middle/src/thir/visit.rs

```diff
@@ -240,36 +240,46 @@ pub fn walk_pat<'thir, 'tcx: 'thir, V: Visitor<'thir, 'tcx>>(
     visitor: &mut V,
     pat: &'thir Pat<'tcx>,
 ) {
-    use PatKind::*;
-    match &pat.kind {
-        AscribeUserType { subpattern, ascription: _ }
-        | Deref { subpattern }
-        | DerefPattern { subpattern, .. }
-        | Binding { subpattern: Some(subpattern), .. } => visitor.visit_pat(subpattern),
-        Binding { .. } | Wild | Never | Error(_) => {}
-        Variant { subpatterns, adt_def: _, args: _, variant_index: _ } | Leaf { subpatterns } => {
-            for subpattern in subpatterns {
-                visitor.visit_pat(&subpattern.pattern);
+    for_each_immediate_subpat(pat, |p| visitor.visit_pat(p));
+}
+
+/// Invokes `callback` on each immediate subpattern of `pat`, if any.
+/// A building block for assembling THIR pattern visitors.
+pub(crate) fn for_each_immediate_subpat<'a, 'tcx>(
+    pat: &'a Pat<'tcx>,
+    mut callback: impl FnMut(&'a Pat<'tcx>),
+) {
+    match pat.kind {
```

Comment:
why did you change this to stop using `&pat.kind`? i think `ref` on every subpattern field is harder to read (and feels like it's not taking advantage of match ergonomics)

Reply:
Ah, the original context was all of the owned subpatterns becoming IDs, so it was useful to be able to bind them by value and not have to add `*` dereferences everywhere.

But you're right that it doesn't make sense for this PR in isolation. I'll change it back to use match ergonomics.